hpct-ldap-server/src/servers/ldap.py

```diff
@@ -0,0 +1,503 @@
+        gid=None,
+        group=None,
+    ) -> None:
+        """Add group.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gid : int
+            Group id.
+        group : str
+            Group name.
+        """
+        dcs = self._split_domain(domain)
+
+        if None not in [domain, gid, group, admin_passwd] and gid > MIN_GID:
+            binddn = f"cn=admin,{dcs}"
+            base_group = [
+                f"dn: cn={group},ou=Groups,{dcs}",
+                "objectClass: posixGroup",
+                f"cn: {group}",
+                f"gidNumber: {gid}",
+            ]
+
+            with tempfile.NamedTemporaryFile(mode="w+t", delete=True) as f:
+                for line in base_group:
+                    f.write(f"{line}\n")
+                    f.flush()
+                cmd = ["ldapadd", "-x", "-D", binddn, "-w", admin_passwd, "-f", f.name]
+
+                rc = subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+                if rc != 0:
+                    raise Exception(
+                        f'Unable to add group. Make sure to run "ldapadd -x -D cn=admin,{dcs} -W -f /etc/ldap/basedn.ldif" first.'
+                    )
+
+    def disable(self) -> None:
+        """Disable services."""
+        for name in self.systemd_services:
+            systemd.service_pause(name)
+
+    def enable(self) -> None:
+        """Enable services."""
+        for name in self.systemd_services:
+            systemd.service_resume(name)
+
+    def install(self) -> None:
+        """Install packages."""
+        if self.packages:
+            try:
+                apt.update()
+                for name in self.packages:
+                    apt.add_package(name)
+            except:
+                raise Exception(f"failed to install package ({name})")
+
+    def is_enabled(self) -> bool:
+        """Check enabled status of services."""
+        if self.systemd_services:
+            for name in self.systemd_services:
+                if not systemd._systemctl("is-enabled", name, quiet=True):
+                    return False
+
+        return True
+
+    def is_installed(self) -> bool:
+        """Check packages are installed."""
+        if self.packages:
+            for name in self.packages:
+                if not apt.DebianPackage.from_installed_package(name).present:
+                    return False
+
+        return True
+
+    def is_running(self) -> bool:
+        """Check running/active status of services."""
+        if self.systemd_services:
+            for name in self.systemd_services:
+                if not systemd.service_running(name):
+                    return False
+
+        return True
+
+    def restart(self) -> None:
+        """Restart servers/services."""
+        self.stop()
+        self.start()
+
+    def set_config(self, admin_passwd=None, domain=None, org=None) -> None:
```

Comment:
Method name from `set_config` -> `configure`? I know the `set_config` matches the action name.